[SQL] cannot find function `cast_to_i8_LongInterval_YEARS`

**Environment:** 

https://try.feldera.com/

**SQL program:** 

CREATE VIEW cast_032 AS SELECT CAST(INTERVAL '1' YEAR AS TINYINT)


**Error message:**

error[E0425]: cannot find function `cast_to_i8_LongInterval_YEARS` in this scope
    --> crates/feldera_pipe_operator_098c5ac7854648a5/src/lib.rs:72:104
     |
  72 | ..., &sourceMap, cast_to_i8_LongInterval_YEARS(LongInterval::from_months(12)))));
     |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `cast_to_VN_LongInterval_YEARS`
     |
    ::: /home/ubuntu/feldera/crates/sqllib/src/casts.rs:3591:1
     |
3591 | cast_variant!(LongInterval_YEARS, LongInterval, LongInterval);
     | ------------------------------------------------------------- similarly named function `cast_to_VN_LongInterval_YEARS` defined here
error: could not compile `feldera_pipe_operator_098c5ac7854648a5` (lib) due to 1 previous error

--- a/crates/sqllib/src/casts.rs
+++ b/crates/sqllib/src/casts.rs
@@ -1739,202 +1739,38 @@ pub fn cast_to_i64_Weight(w: Weight) -> SqlResult<i64> {
 }
 
 #[doc(hidden)]
-pub fn cast_to_i8_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<i8> {
-    cast_to_i8_i64(value.milliseconds())
+pub fn cast_to_i8_LongInterval_YEARS(value: LongInterval) -> SqlResult<i8> {
+    cast_to_i8_i32(value.years())
 }
 
 #[doc(hidden)]
-pub fn cast_to_i16_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<i16> {
-    cast_to_i16_i64(value.milliseconds())
+pub fn cast_to_i8_LongInterval_MONTHS(value: LongInterval) -> SqlResult<i8> {
+    cast_to_i8_i32(value.months())
 }
 
 #[doc(hidden)]
-pub fn cast_to_i32_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<i32> {
-    cast_to_i32_i64(value.milliseconds())
+pub fn cast_to_i16_LongInterval_YEARS(value: LongInterval) -> SqlResult<i16> {
+    cast_to_i16_i32(value.years())
 }
 
 #[doc(hidden)]
-pub fn cast_to_i64_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<i64> {
-    Ok(value.milliseconds())
+pub fn cast_to_i16_LongInterval_MONTHS(value: LongInterval) -> SqlResult<i16> {
+    cast_to_i16_i32(value.months())
 }
 
 #[doc(hidden)]
-pub fn cast_to_i8_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<i8> {
-    cast_to_i8_i64(value.milliseconds())
+pub fn cast_to_i32_LongInterval_YEARS(value: LongInterval) -> SqlResult<i32> {
+    Ok(value.years())
 }
 
 #[doc(hidden)]
-pub fn cast_to_i16_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<i16> {
-    cast_to_i16_i64(value.milliseconds())
+pub fn cast_to_i32_LongInterval_MONTHS(value: LongInterval) -> SqlResult<i32> {
+    Ok(value.months())
 }
-
-#[doc(hidden)]
-pub fn cast_to_i32_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<i32> {
-    cast_to_i32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i64_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<i64> {
-    Ok(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i8_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<i8> {
-    cast_to_i8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i16_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<i16> {
-    cast_to_i16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i32_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<i32> {
-    cast_to_i32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i64_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<i64> {
-    Ok(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i8_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<i8> {
-    cast_to_i8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i16_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<i16> {
-    cast_to_i16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i32_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<i32> {
-    cast_to_i32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_i64_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<i64> {
-    Ok(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u8_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<u8> {
-    cast_to_u8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u16_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<u16> {
-    cast_to_u16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u32_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<u32> {
-    cast_to_u32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u64_ShortInterval_DAYS(value: ShortInterval) -> SqlResult<u64> {
-    cast_to_u64_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u8_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<u8> {
-    cast_to_u8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u16_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<u16> {
-    cast_to_u16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u32_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<u32> {
-    cast_to_u32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u64_ShortInterval_HOURS(value: ShortInterval) -> SqlResult<u64> {
-    cast_to_u64_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u8_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<u8> {
-    cast_to_u8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u16_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<u16> {
-    cast_to_u16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u32_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<u32> {
-    cast_to_u32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u64_ShortInterval_MINUTES(value: ShortInterval) -> SqlResult<u64> {
-    cast_to_u64_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u8_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<u8> {
-    cast_to_u8_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u16_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<u16> {
-    cast_to_u16_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u32_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<u32> {
-    cast_to_u32_i64(value.milliseconds())
-}
-
-#[doc(hidden)]
-pub fn cast_to_u64_ShortInterval_SECONDS(value: ShortInterval) -> SqlResult<u64> {
-    cast_to_u64_i64(value.milliseconds())
-}
-
-cast_function!(i8, i8, ShortInterval_DAYS, ShortInterval);
-cast_function!(i16, i16, ShortInterval_DAYS, ShortInterval);
-cast_function!(i32, i32, ShortInterval_DAYS, ShortInterval);
-cast_function!(i64, i64, ShortInterval_DAYS, ShortInterval);
-cast_function!(i8, i8, ShortInterval_HOURS, ShortInterval);
-cast_function!(i16, i16, ShortInterval_HOURS, ShortInterval);
-cast_function!(i32, i32, ShortInterval_HOURS, ShortInterval);
-cast_function!(i64, i64, ShortInterval_HOURS, ShortInterval);
-cast_function!(i8, i8, ShortInterval_MINUTES, ShortInterval);
-cast_function!(i16, i16, ShortInterval_MINUTES, ShortInterval);
-cast_function!(i32, i32, ShortInterval_MINUTES, ShortInterval);
-cast_function!(i64, i64, ShortInterval_MINUTES, ShortInterval);
-cast_function!(i8, i8, ShortInterval_SECONDS, ShortInterval);
-cast_function!(i16, i16, ShortInterval_SECONDS, ShortInterval);
-cast_function!(i32, i32, ShortInterval_SECONDS, ShortInterval);
-cast_function!(i64, i64, ShortInterval_SECONDS, ShortInterval);
-
-cast_function!(u8, u8, ShortInterval_DAYS, ShortInterval);
-cast_function!(u16, u16, ShortInterval_DAYS, ShortInterval);
-cast_function!(u32, u32, ShortInterval_DAYS, ShortInterval);
-cast_function!(u64, u64, ShortInterval_DAYS, ShortInterval);
-cast_function!(u8, u8, ShortInterval_HOURS, ShortInterval);
-cast_function!(u16, u16, ShortInterval_HOURS, ShortInterval);
-cast_function!(u32, u32, ShortInterval_HOURS, ShortInterval);
-cast_function!(u64, u64, ShortInterval_HOURS, ShortInterval);
-cast_function!(u8, u8, ShortInterval_MINUTES, ShortInterval);
-cast_function!(u16, u16, ShortInterval_MINUTES, ShortInterval);
-cast_function!(u32, u32, ShortInterval_MINUTES, ShortInterval);
-cast_function!(u64, u64, ShortInterval_MINUTES, ShortInterval);
-cast_function!(u8, u8, ShortInterval_SECONDS, ShortInterval);
-cast_function!(u16, u16, ShortInterval_SECONDS, ShortInterval);
-cast_function!(u32, u32, ShortInterval_SECONDS, ShortInterval);
-cast_function!(u64, u64, ShortInterval_SECONDS, ShortInterval);
 
 #[doc(hidden)]
 pub fn cast_to_i64_LongInterval_YEARS(value: LongInterval) -> SqlResult<i64> {
-    Ok(value.months() as i64)
+    Ok(value.years() as i64)
 }
 
 #[doc(hidden)]
@@ -1942,8 +1778,63 @@ pub fn cast_to_i64_LongInterval_MONTHS(value: LongInterval) -> SqlResult<i64> {
     Ok(value.months() as i64)
 }
 
+#[doc(hidden)]
+pub fn cast_to_u8_LongInterval_YEARS(value: LongInterval) -> SqlResult<u8> {
+    cast_to_u8_i32(value.years())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u8_LongInterval_MONTHS(value: LongInterval) -> SqlResult<u8> {
+    cast_to_u8_i32(value.months())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u16_LongInterval_YEARS(value: LongInterval) -> SqlResult<u16> {
+    cast_to_u16_i32(value.years())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u16_LongInterval_MONTHS(value: LongInterval) -> SqlResult<u16> {
+    cast_to_u16_i32(value.months())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u32_LongInterval_YEARS(value: LongInterval) -> SqlResult<u32> {
+    cast_to_u32_i32(value.years())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u32_LongInterval_MONTHS(value: LongInterval) -> SqlResult<u32> {
+    cast_to_u32_i32(value.months())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u64_LongInterval_YEARS(value: LongInterval) -> SqlResult<u64> {
+    cast_to_u64_i32(value.years())
+}
+
+#[doc(hidden)]
+pub fn cast_to_u64_LongInterval_MONTHS(value: LongInterval) -> SqlResult<u64> {
+    cast_to_u64_i32(value.months())
+}
+
+cast_function!(i8, i8, LongInterval_MONTHS, LongInterval);
+cast_function!(i8, i8, LongInterval_YEARS, LongInterval);
+cast_function!(i16, i16, LongInterval_MONTHS, LongInterval);
+cast_function!(i16, i16, LongInterval_YEARS, LongInterval);
+cast_function!(i32, i32, LongInterval_MONTHS, LongInterval);
+cast_function!(i32, i32, LongInterval_YEARS, LongInterval);
 cast_function!(i64, i64, LongInterval_MONTHS, LongInterval);
 cast_function!(i64, i64, LongInterval_YEARS, LongInterval);
+
+cast_function!(u8, u8, LongInterval_MONTHS, LongInterval);
+cast_function!(u8, u8, LongInterval_YEARS, LongInterval);
+cast_function!(u16, u16, LongInterval_MONTHS, LongInterval);
+cast_function!(u16, u16, LongInterval_YEARS, LongInterval);
+cast_function!(u32, u32, LongInterval_MONTHS, LongInterval);
+cast_function!(u32, u32, LongInterval_YEARS, LongInterval);
+cast_function!(u64, u64, LongInterval_MONTHS, LongInterval);
+cast_function!(u64, u64, LongInterval_YEARS, LongInterval);
 
 //////// casts to Short interval
 

--- a/crates/sqllib/src/interval.rs
+++ b/crates/sqllib/src/interval.rs
@@ -54,6 +54,13 @@ pub struct ShortInterval {
 }
 
 #[doc(hidden)]
+pub fn reinterpret_ShortInterval(value: ShortInterval) -> i64 {
+    value.milliseconds()
+}
+
+some_polymorphic_function1!(reinterpret, ShortInterval, ShortInterval, i64);
+
+#[doc(hidden)]
 impl<D> ::rkyv::Deserialize<ShortInterval, D> for ArchivedShortInterval
 where
     D: ::rkyv::Fallible + ::core::any::Any,
@@ -484,6 +491,13 @@ impl LongInterval {
         (months / 12) * mul
     }
 }
+
+#[doc(hidden)]
+pub fn reinterpret_LongInterval(value: LongInterval) -> i64 {
+    value.months() as i64
+}
+
+some_polymorphic_function1!(reinterpret, LongInterval, LongInterval, i64);
 
 #[doc(hidden)]
 pub fn abs_LongInterval(value: LongInterval) -> LongInterval {

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -1876,6 +1876,17 @@ public class ToRustInnerVisitor extends InnerVisitor {
             this.builder.append(")");
             this.pop(expression);
             return VisitDecision.STOP;
+        } else if (expression.opcode == DBSPOpcode.REINTERPRET) {
+            DBSPTypeBaseType sourceType = expression.source.getType().to(DBSPTypeBaseType.class);
+            this.builder.append(expression.opcode.toString())
+                    .append("_")
+                    .append(sourceType.shortName())
+                    .append(sourceType.nullableSuffix())
+                    .append("(");
+            expression.source.accept(this);
+            this.builder.append(")");
+            this.pop(expression);
+            return VisitDecision.STOP;
         }
         if (expression.source.getType().mayBeNull) {
             this.builder.append("(")

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -41,7 +41,6 @@ import org.apache.calcite.sql.SqlKind;
 import org.apache.calcite.util.DateString;
 import org.apache.calcite.util.TimeString;
 import org.apache.calcite.util.TimestampString;
-import org.dbsp.sqlCompiler.circuit.operator.DBSPConcreteAsofJoinOperator;
 import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.ICompilerComponent;
 import org.dbsp.sqlCompiler.compiler.errors.BaseCompilerException;
@@ -1061,9 +1060,15 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                 return makeBinaryExpressions(node, type, DBSPOpcode.BW_OR, ops);
             case BIT_XOR:
                 return makeBinaryExpressions(node, type, DBSPOpcode.XOR, ops);
+            case REINTERPRET:
+                // Calcite only seems to generate this for converting intervals to integers
+                if ((!type.is(DBSPTypeInteger.class) || type.to(DBSPTypeInteger.class).getWidth() != 64)
+                        && !ops.get(0).getType().is(IsIntervalType.class)) {
+                    throw new InternalCompilerError("Unexpected cast ", node);
+                }
+                return new DBSPUnaryExpression(node, type, DBSPOpcode.REINTERPRET, ops.get(0));
             case CAST:
             case SAFE_CAST:
-            case REINTERPRET:
                 if (!validateCast(ops.get(0).getType(), type, call.op.kind == SqlKind.SAFE_CAST)) {
                     throw new CompilationError(call.op.kind + " cannot be used to convert " +
                             ops.get(0).getType().asSqlString() +

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
@@ -100,7 +100,9 @@ public enum DBSPOpcode {
     // Same as MAP_CONVERT, except the function applied returns SqlResult, and
     // the map function returns None in case any partial result is Error.
     MAP_CONVERT_SAFE("map_map_safe", false),
-    ;
+    // Used to implement Calcite's REINTERPRET casts;
+    // input is always an interval, output is always i64
+    REINTERPRET("reinterpret", false);
 
     private final String text;
     public final boolean isAggregate;

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPUnaryExpression.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPUnaryExpression.java
@@ -79,7 +79,8 @@ public final class DBSPUnaryExpression extends DBSPExpression {
         if (this.opcode == DBSPOpcode.DECIMAL_TO_INTEGER ||
                 this.opcode == DBSPOpcode.INTEGER_TO_DECIMAL ||
                 this.opcode == DBSPOpcode.SHORT_INTERVAL_TO_INTEGER ||
-                this.opcode == DBSPOpcode.INTEGER_TO_SHORT_INTERVAL) {
+                this.opcode == DBSPOpcode.INTEGER_TO_SHORT_INTERVAL ||
+                this.opcode == DBSPOpcode.REINTERPRET) {
             return builder.append(this.opcode.toString())
                     .append("(")
                     .append(this.source)

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
@@ -23,6 +23,7 @@
 
 package org.dbsp.sqlCompiler.compiler.sql.simple;
 
+import org.dbsp.sqlCompiler.compiler.CompilerOptions;
 import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 import org.dbsp.sqlCompiler.compiler.sql.tools.Change;
@@ -636,8 +637,8 @@ public class CastTests extends SqlIoTest {
         Assert.assertEquals(types.length, legal.length);
         Assert.assertEquals(types.length, legal[0].length);
         StringBuilder program = new StringBuilder();
-        program.append("CREATE VIEW V AS SELECT ");
         boolean first = true;
+        program.append("CREATE VIEW V AS SELECT ");
         for (int i = 0; i < types.length; i++) {
             String type = types[i];
             String value = values[i];
@@ -684,6 +685,55 @@ public class CastTests extends SqlIoTest {
             }
             program.append(";\n");
         }
-        this.compileRustTestCase(program.toString());
+        // Disable Calcite optimizations so it doesn't do constant-folding
+        CompilerOptions options = this.testOptions();
+        options.languageOptions.optimizationLevel = 0;
+        DBSPCompiler compiler = new DBSPCompiler(options);
+        compiler.submitStatementsForCompilation(program.toString());
+        this.getCCS(compiler);
+    }
+
+    @Test
+    public void issue5894() {
+        this.qs("""
+                SELECT INTERVAL '1' YEAR;
+                 r
+                ---
+                 1 year
+                (1 row)
+                
+                SELECT CAST('1' AS INTERVAL YEAR);
+                 r
+                ---
+                 1 year
+                (1 row)
+                
+                SELECT INTERVAL 1 YEAR;
+                 r
+                ---
+                 1 year
+                (1 row)
+                
+                SELECT CAST(CAST('1' AS INTERVAL YEAR) AS TINYINT);
+                 r
+                ---
+                 1
+                (1 row)
+                
+                SELECT CAST(INTERVAL '1' YEAR AS TINYINT);
+                 r
+                ---
+                 1
+                (1 row)""");
+    }
+
+    @Test
+    public void issue5895() {
+        this.qs("""
+                SELECT CAST(INTERVAL '1000' MONTH AS INT);
+                 r
+                -----
+                 1000
+                (1 row)""");
     }
 }

